src/XMonad/Operations.hs

```diff
@@ -500,17 +500,20 @@ isClient :: Window -> X Bool
 isClient w = withWindowSet $ return . W.member w
 
 -- | Combinations of extra modifier masks we need to grab keys\/buttons for.
--- (numlock and capslock)
+-- (by default numlock and capslock, can be overridden in 'stripModMask')
 extraModifiers :: X [KeyMask]
 extraModifiers = do
-    nlm <- gets numberlockMask
-    return [0, nlm, lockMask, nlm .|. lockMask ]
+    smm <- join $ asks $ stripModMask . config
+    return $ map (foldl' (.|.) 0) (subsequences (bits smm))
+  where
+    bits 0 = []
+    bits n = let b = countTrailingZeros n in bit b : bits (n `clearBit` b)
```

Comment:
This would be much more understandable if we just specified a list in `stripModMask`, right?[^1]  Is there any reason why we don't do that and instead or everything together there already?

[^1]: AFAICT we could then just use `subsequences` alone.

Reply:
"Understandable" refers to the implementation or the user interface? I wanted to abstract away the implementation detail that we need to grab all bit combinations and just expose a _mask_ that's directly used in `cleanMask` and processed into combinations in here.

You could argue, however, that abstracting the implementation detail is harmful in here, because people who want to not strip numlock will need to manually grab some keys with and without numlock. Hm, not sure what to do. :-/

Reply:
I would argue that the fact that key masks are or'd together is an implementation detail in itself, which is not necessarily known to the user (many people don't know what the funnly looking `.|.` is doing in their configs :).  Intuitively, I would expect an option along the lines of "filter out these modifiers" to take a list of key masks instead of a single one